frontend/app/components/wp-edit-form/work-package-changeset.ts

```diff
@@ -0,0 +1,286 @@
+  public schemaCacheService:SchemaCacheService;
+  public wpCacheService:WorkPackageCacheService;
+  public wpCreate:WorkPackageCreateService;
+
+  // The changeset to be applied to the work package
+  private changes:{[attribute:string]:any} = {};
+  public inFlight:boolean = false;
+
+  // The current work package form
+  public wpForm = input<FormResourceInterface>();
+
+  // The current editing resource state
+  public resource = input<HalResource>();
+
+  constructor(public workPackage:WorkPackageResourceInterface) {
+    $injectFields(
+      this, 'wpNotificationsService', '$q', 'schemaCacheService',
+      'wpCacheService', 'wpCreate'
+    );
+
+    if (this.workPackage.isNew) {
+      // New work packages have no schema set yet, so update the form immediately to get one
+      this.updateForm();
+    } else {
+      // Start with a resource from the current work package knowledge.
+      const payload = this.mergeWithPayload(workPackage.$plain);
+      this.buildResource(payload);
+    }
+  }
+
+  public startEditing(key:string) {
+    if (!this.isOverridden(key)) {
+      this.changes[key] = _.cloneDeep(this.workPackage[key]);
+    }
+  }
+
+  public reset(key:string) {
+    delete this.changes[key];
+  }
+
+  public clear() {
+    this.changes = {};
+  }
+
+  public get empty() {
+    return _.isEmpty(this.changes);
+  }
+
+  /**
+   * Retrieve the editing value for the given attribute
+   *
+   * @param {string} key The attribute to read
+   * @return {any} Either the value from the overriden change, or the default value
+   */
+  public value(key:string) {
+    if (this.isOverridden(key)) {
+      return this.changes[key];
+    } else {
+      return this.workPackage[key];
+    }
+  }
+
+  public setValue(key:string, val:any) {
+    this.changes[key] = val;
+
+    // Update the form for fields that may alter the form itself.
+    if (key === 'project' || key === 'type') {
+      this.updateForm();
+    }
+  }
+
+  /**
+   * Return whether a change value exist for the given attribute key.
+   * @param {string} key
+   * @return {boolean}
+   */
+  public isOverridden(key:string) {
+    return this.changes.hasOwnProperty(key);
+  }
+
+  public getForm():Promise<FormResourceInterface> {
+    this.wpForm.putFromPromiseIfPristine(() => this.updateForm());
+
+    if (this.wpForm.hasValue()) {
+      return Promise.resolve(this.wpForm.value);
+    } else {
+      return new Promise((resolve,) => this.wpForm.valuesPromise().then(resolve));
+    }
+  }
+
+  /**
+   * Update the form resource from the API.
+   * @return {angular.IPromise<any>}
+   */
+  public updateForm():ng.IPromise<FormResourceInterface> {
+    // Always resolve form to the latest form
+    // This way, we won't have to actively reset it.
+    // But store the existing form in case of an error.
+    // Because if we get an error, the object returned is not a form
+    // and thus lacks the links the implementation depends upon.
+    const oldForm = this.wpForm.value;
```

Comment:
If I get this right, we do no longer alter the work package. As such, we should no longer have to rely on links in the form. Do we still need to juggle the old/new form?